demonstrations/tutorial_vqt.py

```diff
@@ -0,0 +1,512 @@
+        rotation.append(np.split(i, 3))
+
+    ansatz_params = [rotation, coupling]
+
+    return [dist_params, ansatz_params]
+
+
+######################################################################
+# We then pass this function, along with the ansatz and the entropy
+# function into the final cost function:
+#
+
+
+def exact_cost(params):
+
+    global iterations
+
+    # Transforms the parameter list
+    parameters = convert_list(params)
+    dist_params = parameters[0]
+    ansatz_params = parameters[1]
+
+    # Creates the probability distribution
+    distribution = prob_dist(dist_params)
+
+    # Generates a list of all computational basis states, of our qubit system
+    combos = itertools.product([0, 1], repeat=N)
+    s = [list(i) for i in combos]
+
+    # Passes each basis state through the variational circuit and multiplies the calculated energy EV with the associated probability from the distribution
+    final_cost = 0
+    for i in s:
+        result = qnode(ansatz_params[0], ansatz_params[1], sample=i)
+        for j in range(0, len(i)):
+            result = result * distribution[j][i[j]]
+        final_cost += result
+
+    # Calculates the entropy and the final cost function
+    entropy = calculate_entropy(distribution)
+    final_final_cost = beta * final_cost - entropy
+
+    if iterations % 20 == 0:
+        print("Cost at Step " + str(iterations) + ": " + str(final_final_cost))
+
+    iterations += 1
+
+    return final_final_cost
+
+
+######################################################################
+# The last step is to define the optimizer, and execute the optimization
+# method. We use the "Constrained Optimization by Linear Approximation" 
+# (`COBYLA <https://en.wikipedia.org/wiki/COBYLA>`__) optimization method, 
+# which a graidient-free optimizer. We observe that for this algorithm, COBYLA 
+# has a lower run-time than it gradient-based counterparts, so we utilize it 
+# in this tutorial:
+#
+
+
+iterations = 0
+
+params = [np.random.randint(-300, 300) / 100 for i in range(0, (N * (1 + depth * 4)))]
+out = minimize(exact_cost, x0=params, method="COBYLA", options={"maxiter": 1600})
+out_params = out["x"]
+print(out)
+
+
+######################################################################
+# We can now check to see how well our optimization method performed by
+# writing a function that re-constructs the transformed density
+# matrix of some initial state, with respect to lists of
+# :math:`\theta` and :math:`\phi` parameters:
+#
+
+
+def prepare_state(params, device):
+
+    # Initializes the density matrix
+
+    final_density_matrix = np.zeros((2 ** N, 2 ** N))
+
+    # Prepares the optimal parameters, creates the distribution and the bitstrings
+    parameters = convert_list(params)
+    dist_params = parameters[0]
+    unitary_params = parameters[1]
+
+    distribution = prob_dist(dist_params)
+
+    combos = itertools.product([0, 1], repeat=N)
+    s = [list(i) for i in combos]
```

Comment:
```suggestion
    s = [list(c) for c in combos]
```
By the way, some of this code is also used in the definition of the cost function. Does it make sense to write a single function that gets used in both places?